apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.js

```diff
@@ -81,256 +95,183 @@ router.post(
-  if (res.locals.question.course_id !== res.locals.course.id) {
-    return next(error.make(403, 'Access denied'));
-  }
-  if (req.body.__action === 'change_id') {
-    debug(`Change qid from ${res.locals.question.qid} to ${req.body.id}`);
-    if (!req.body.id) return next(new Error(`Invalid QID (was falsy): ${req.body.id}`));
-    if (!/^[-A-Za-z0-9_/]+$/.test(req.body.id)) {
-      return next(
-        new Error(
-          `Invalid QID (was not only letters, numbers, dashes, slashes, and underscores, with no spaces): ${req.body.id}`,
-        ),
-      );
-    }
-    let qid_new;
-    try {
-      qid_new = path.normalize(req.body.id);
-    } catch (err) {
-      return next(new Error(`Invalid QID (could not be normalized): ${req.body.id}`));
+router.post(
+  '/',
+  asyncHandler(async (req, res, next) => {
+    if (res.locals.question.course_id !== res.locals.course.id) {
+      throw error.make(403, 'Access denied');
     }
-    if (res.locals.question.qid === qid_new) {
-      debug('The new qid is the same as the old qid - do nothing');
-      res.redirect(req.originalUrl);
-    } else {
-      const editor = new QuestionRenameEditor({
-        locals: res.locals,
-        qid_new: qid_new,
-      });
-      editor.canEdit((err) => {
-        if (ERR(err, next)) return;
-        editor.doEdit((err, job_sequence_id) => {
-          if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-            res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-          } else {
-            res.redirect(req.originalUrl);
-          }
+    if (req.body.__action === 'change_id') {
+      if (!req.body.id) throw new Error(`Invalid QID (was falsy): ${req.body.id}`);
+      if (!/^[-A-Za-z0-9_/]+$/.test(req.body.id)) {
+        throw new Error(
+          `Invalid QID (was not only letters, numbers, dashes, slashes, and underscores, with no spaces): ${req.body.id}`,
+        );
+      }
+      let qid_new;
+      try {
+        qid_new = path.normalize(req.body.id);
+      } catch (err) {
+        throw new Error(`Invalid QID (could not be normalized): ${req.body.id}`);
+      }
+      if (res.locals.question.qid === qid_new) {
+        res.redirect(req.originalUrl);
+      } else {
+        const editor = new QuestionRenameEditor({
+          locals: res.locals,
+          qid_new: qid_new,
         });
-      });
-    }
-  } else if (req.body.__action === 'copy_question') {
-    debug('Copy question');
-    if (idsEqual(req.body.to_course_id, res.locals.course.id)) {
-      // In this case, we are making a duplicate of this question in the same course
-      const editor = new QuestionCopyEditor({
-        locals: res.locals,
-      });
-      editor.canEdit((err) => {
-        if (ERR(err, next)) return;
-        editor.doEdit((err, job_sequence_id) => {
-          if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-            res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
```

Comment:
It looks like I lost some error handling here (and elsewhere where I updated to `doEditAsync`). Is this something I should retain?

Reply:
Yep, we do want to redirect to the `edit_error` page if there was an error. I think this is going to be moderately tricky to get right.

Changing `doEdit` to get rid of `async.series` actually changed how even the callback-based version of the function behaves, and it actually changes it for `doEditAsync()` even more. Before, even if an error was encountered, the callback would receive both the error _and_ the job sequence ID, which is necessary to actually perform the redirect. Now, however, it won't, meaning that all the existing usages of `doEdit` won't behave correctly when an error is encountered.

`doEditAsync` suffers from the same problem: the function can either throw an error, or completely successfully and return the job sequence ID, but it can't do both (that is, we can't get the job sequence ID if it errors).

I can think of at least two ways to solve this problem:

- Throw a custom error class (e.g. `EditorError`) that contains the job sequence ID. When doing an edit, error handling would then look like this:

```ts
try {
  await editor.doEditAsync();
} catch (err) {
  if (err instanceof EditorError) {
    res.redirect(`.../${err.jobSequenceId`);
  } else {
    throw err;
  }
}
```

- Instead of throwing an error to indicate a failed sync, return a value with both the job sequence ID and a status field:
  - If the job succeeded, return something like `{ success: true, jobSequenceId: 12345 }`
  - If the job failed, return something like `{ success: false, jobSequenceId: 12345 }`

I'm going to raise this for discussion on Slack, since I don't know that we have a convention for this yet!

Reply:
After discussing in the dev meeting, we have decided to implement the two function approach. I have split this out to two async functions `setupEditAsync()` which will return the server job, and then `doEditAsync` which accepts the server job as an argument and performs the edit. I have also added a `try{...}/catch{...}` wrapper to `doEditAsync()` which will redirect to the error page if there is an error.